transparency/boot_entry.go

```diff
@@ -0,0 +1,220 @@
+// to validate the boot entry artifacts.
+// Returns error if the boot artifacts are not passing the validation.
+func (b *BootEntry) Validate(c *Config) (err error) {
+	if c.Status == None {
+		return
+	}
+
+	if b == nil || len(*b) == 0 {
+		return fmt.Errorf("got an invalid boot entry pointer")
+	}
+
+	// Automatically load the configuration from the UEFI partition
+	// when the function is used within the UEFI boot loader.
+	if c.UefiRoot != nil {
+		entryPath, err := c.Path(b)
+		if err != nil {
+			return fmt.Errorf("cannot load boot transparency configuration, %v", err)
+		}
+
+		if err = c.loadFromUefiRoot(entryPath); err != nil {
+			return fmt.Errorf("cannot load boot transparency configuration, %v", err)
+		}
+	}
+
+	te, err := transparency.GetEngine(transparency.Sigsum)
+	if err != nil {
+		return fmt.Errorf("unable to configure the transparency engine, %w", err)
+	}
+
+	if err = te.SetKey([]string{string(c.LogKey)}, []string{string(c.SubmitKey)}); err != nil {
+		return fmt.Errorf("unable to set log and submitter keys, %v", err)
+	}
+
+	wp, err := te.ParseWitnessPolicy(c.WitnessPolicy)
+	if err != nil {
+		return fmt.Errorf("unable to parse witness policy, %v", err)
+	}
+
+	if err = te.SetWitnessPolicy(wp); err != nil {
+		return fmt.Errorf("unable to set witness policy, %v", err)
+	}
+
+	pb, _, err := te.ParseProof(c.ProofBundle)
+	if err != nil {
+		return fmt.Errorf("unable to parse proof bundle, %v", err)
+	}
+
+	// If network access is available the inclusion proof verification
+	// is performed using the proof fetched from the log.
+	if c.Status == Online {
+		pr, err := te.GetProof(pb)
+		if err != nil {
+			return err
+		}
+
+		freshBundle := pb.(*sigsum.ProofBundle)
+		freshBundle.Proof = string(pr)
+
+		if err = te.VerifyProof(freshBundle); err != nil {
+			return err
+		}
+	} else {
+		// If network access is not available the inclusion proof verification
+		// is performed using the proof included in the proof bundle.
+		if err = te.VerifyProof(pb); err != nil {
+			return err
+		}
+	}
+
+	requirements, err := policy.ParseRequirements(c.BootPolicy)
+	if err != nil {
+		return
+	}
+
+	proof := pb.(*sigsum.ProofBundle)
+
+	claims, err := policy.ParseStatement(proof.Statement)
+	if err != nil {
+		return
+	}
+
+	// Validate the matching between loaded artifact hashes and
+	// the ones included in the proof bundle.
+	if err = b.validateProofHashes(claims); err != nil {
+		return
+	}
+
+	// Validate the matching between the logged claims and the policy requirements.
+	if err = policy.Validate(requirements, claims); err != nil {
+		// The boot bundle is NOT authorized.
+		return
+	}
+
+	// boot-transparency validation passed, boot bundle is authorized.
+	return
+}
+
+// Hash performs data hashing using SHA-256, which is
+// the same algorithm used by boot-transparency library.
+// Returns the computed hash as hex string.
+func Hash(data *[]byte) (hexHash string) {
+	h := sha256.New()
```

Comment:
The hash type should come from a constant in boot-transparency rather than being implicit knowledge here, if hash format is mandated by boot-transparency then primitives for hashing belong there.